From b42fe4c58b09b119fe918736be5100ef21b7b9cd Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Mon, 18 Apr 2016 17:34:50 +0300
Subject: [PATCH 357/538] vfio: xhci: add support for XHCI reset hook

The VFIO reset hook is called every time a platform device is
passed to a guest or removed from a guest.
When the XHCI device is unbound from the host, the host driver
disables the XHCI clock so when the device is passed to the
guest it becomes dis-functional.
This initial implementation uses the VFIO reset hook to enable
the XHCI clocks on behalf of the guest.

Change-Id: I1e729645b9cb600df39b5bd8ff213ec34a137d2d
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29163
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Shadi Ammouri <shadi@marvell.com>
---
 drivers/vfio/platform/reset/Kconfig              |  8 ++++
 drivers/vfio/platform/reset/Makefile             |  2 +
 drivers/vfio/platform/reset/vfio_platform_xhci.c | 59 ++++++++++++++++++++++++
 3 files changed, 69 insertions(+)
 create mode 100644 drivers/vfio/platform/reset/vfio_platform_xhci.c

diff --git a/drivers/vfio/platform/reset/Kconfig b/drivers/vfio/platform/reset/Kconfig
index 70cccc5..0010abc 100644
--- a/drivers/vfio/platform/reset/Kconfig
+++ b/drivers/vfio/platform/reset/Kconfig
@@ -6,6 +6,14 @@ config VFIO_PLATFORM_CALXEDAXGMAC_RESET
 
 	  If you don't know what to do here, say N.
 
+config VFIO_PLATFORM_XHCI_RESET
+	tristate "VFIO support for USB XHCI reset"
+	depends on VFIO_PLATFORM
+	help
+	  Enables the VFIO platform driver to handle reset for USB XHCI
+
+	  If you don't know what to do here, say N.
+
 config VFIO_PLATFORM_AMDXGBE_RESET
 	tristate "VFIO support for AMD XGBE reset"
 	depends on VFIO_PLATFORM
diff --git a/drivers/vfio/platform/reset/Makefile b/drivers/vfio/platform/reset/Makefile
index 93f4e23..40b7f60 100644
--- a/drivers/vfio/platform/reset/Makefile
+++ b/drivers/vfio/platform/reset/Makefile
@@ -1,7 +1,9 @@
 vfio-platform-calxedaxgmac-y := vfio_platform_calxedaxgmac.o
+vfio-platform-xhci-y := vfio_platform_xhci.o
 vfio-platform-amdxgbe-y := vfio_platform_amdxgbe.o
 
 ccflags-y += -Idrivers/vfio/platform
 
 obj-$(CONFIG_VFIO_PLATFORM_CALXEDAXGMAC_RESET) += vfio-platform-calxedaxgmac.o
+obj-$(CONFIG_VFIO_PLATFORM_XHCI_RESET) += vfio-platform-xhci.o
 obj-$(CONFIG_VFIO_PLATFORM_AMDXGBE_RESET) += vfio-platform-amdxgbe.o
diff --git a/drivers/vfio/platform/reset/vfio_platform_xhci.c b/drivers/vfio/platform/reset/vfio_platform_xhci.c
new file mode 100644
index 0000000..f1516f3
--- /dev/null
+++ b/drivers/vfio/platform/reset/vfio_platform_xhci.c
@@ -0,0 +1,59 @@
+/*
+ * VFIO platform driver specialized for XHCI reset
+ * reset code is inherited from XHCI native driver
+ *
+ * Copyright 2016 Marvell Semiconductors, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/io.h>
+#include <linux/device.h>
+#include <linux/of.h>
+#include <linux/clk.h>
+
+#include "vfio_platform_private.h"
+
+#define DRIVER_VERSION  "0.1"
+#define DRIVER_AUTHOR   "Yehuda Yitschak <yehuday@marvell.com>"
+#define DRIVER_DESC     "Reset support for XHCI vfio platform device"
+
+#define MAX_XHCI_CLOCKS		4
+
+int vfio_platform_xhci_reset(struct vfio_platform_device *vdev)
+{
+	struct device_node *np = vdev->device->of_node;
+	struct clk *clk;
+	int ret, i;
+
+	for (i = 0; i < MAX_XHCI_CLOCKS; i++) {
+		clk = of_clk_get(np, i);
+		if (!IS_ERR(clk)) {
+			ret = clk_prepare_enable(clk);
+			if (ret)
+				return -ENODEV;
+		}
+	}
+
+	return 0;
+}
+
+module_vfio_reset_handler("generic-xhci", vfio_platform_xhci_reset);
+
+MODULE_VERSION(DRIVER_VERSION);
+MODULE_LICENSE("GPL v2");
+MODULE_AUTHOR(DRIVER_AUTHOR);
+MODULE_DESCRIPTION(DRIVER_DESC);
-- 
1.9.1

